quodlibet/quodlibet/formats/_audio.py

```diff
@@ -536,24 +538,104 @@ def _role_call(self, role_tag, sub_keys=None):
-                  .replace(u'/', u'')[:128])
-
-        if os.name == "nt":
-            # this was added at a later point. only use escape_filename here
-            # to keep the linux case the same as before
-            filename = escape_filename(filename)
-            sub_dir = escape_filename(sub_dir)
-        else:
-            filename = fsnative(filename)
-            sub_dir = fsnative(sub_dir)
-
-        path = os.path.join(
-            expanduser(fsnative(u"~/.lyrics")), sub_dir, filename)
-        return path
+        """Returns the validated, or default, lyrics filename for this
+        file. User defined '[memory] lyric_rootpaths' and
+        '[memory] lyric_filenames' matches take precedence"""
+
+        from quodlibet.pattern import ArbitraryExtensionFileFromPattern
+
+        def sanitise(sep, parts):
+            return sep.join(list(map(lambda s: s.replace(u'/', u'')[:128],
+                                     parts)))
+
+        # setup defaults (user-defined take precedence)
+        # root search paths
+        lyric_paths = \
+            config.getstringlist("memory", "lyric_rootpaths", [])
+        # ensure default paths
+        lyric_paths.append(os.path.join(get_home_dir(), ".lyrics"))
+        lyric_paths.append(
+            os.path.join(os.path.dirname(self.comma('~filename'))))
+        # search pathfile names
+        lyric_filenames = \
+            config.getstringlist("memory", "lyric_filenames", [])
+        # ensure some default pathfile names
+        lyric_filenames.append(
+            sanitise(os.sep, [(self.comma("lyricist") or
+                              self.comma("artist")),
+                              self.comma("title")]) + u'.lyric')
+        lyric_filenames.append(
+            sanitise(' - ', [(self.comma("lyricist") or
+                             self.comma("artist")),
+                             self.comma("title")]) + u'.lyric')
+
+        # generate all potential paths (unresolved/unexpanded)
+        pathfiles = OrderedDict()
+        for p in lyric_paths:
+            for f in lyric_filenames:
+                pathfile = ""
+                if os.name == "nt":
+                    # this was added at a later point. only use
+                    # escape_filename here to keep the linux case
+                    # the same as before
+                    pathfile = \
+                        os.path.join(p, os.path.dirname(f),
+                                     escape_filename(os.path.basename(f)))
+                else:
+                    pathfile = \
+                        os.path.join(p, os.path.dirname(f),
+                                    fsnative(os.path.basename(f)))
+                if not pathfile in pathfiles:
+                    pathfiles[pathfile] = pathfile
+
+        #print_d("searching for lyrics in:\n%s" % '\n'.join(pathfiles.keys()))
+
+        # expand each raw pathfile in turn and test for existence
```

Comment:
This could be a function - any time there's a block with a comment we can normally extract to function (then use docstrings :) )

Reply:
done